File: android/app/src/main/java/com/clovis/app/dossiers/DossiersPlugin.kt
// Cree le 25/08/2026, Bourama : Lot 3B Partie 3 mobile (fusion Capacitor).
// Porte depuis clovis-mobile/android-legacy-natif (Lot 2, DossiersScreen.kt
// + DossiersDesignesRepository.kt) : meme logique SAF, l'UI (liste/navigation/
// dialogues) est retiree ici, a construire cote clovis-frontend (web) qui
// appelle ce plugin, voir echange avec Bourama (25/08, Lot 3B).
//
// Cote JS :
//   import { registerPlugin } from '@capacitor/core';
//   const Dossiers = registerPlugin<any>('Dossiers');
//   await Dossiers.choisirDossier();               // ouvre le selecteur systeme
//   await Dossiers.listerDossiersDesignes();
//   await Dossiers.listerContenu({ uri });
//   await Dossiers.lireFichier({ uri });            // ajoute 30/08/2026, contenu brut base64
//   await Dossiers.creerSousDossier({ parentUri, nom });
//   await Dossiers.creerFichier({ parentUri, nom, typeMime });
//   await Dossiers.renommer({ elementUri, nouveauNom });
//   await Dossiers.supprimer({ elementUri });
//   await Dossiers.deplacer({ elementUri, ancienParentUri, nouveauParentUri });
//   await Dossiers.retirerDossierDesigne({ uri });
//
// Ajoute le 26/08/2026 : chaque changement (ajout/retrait) et l'ouverture
// de l'app (load()) declenchent maintenant automatiquement une
// synchronisation silencieuse avec le backend (miroir des NOMS de
// dossiers designes, voir clovis-backend/core/dossiers_designes_mobile.py)
// (android/ et ios/ là-bas), c'est transparent.
package com.clovis.app.dossiers

import android.content.Intent
import android.net.Uri
import android.util.Log
import com.clovis.app.pont.ClovisApiClient
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

@CapacitorPlugin(name = "Dossiers")
class DossiersPlugin : Plugin() {

    private val repo by lazy { DossiersDesignesRepository(context) }

    // Ajoute le 26/08/2026 : miroir cote backend (voir
    // clovis-backend/core/dossiers_designes_mobile.py) pour que l'agent
    // sache quels noms de dossiers cibler. Appele a l'ouverture de l'app
    // (load()) et apres CHAQUE changement (ajout/retrait), jamais
    // bloquant pour l'utilisateur : echec silencieux (log seulement), la
    // prochaine synchronisation (ouverture suivante ou prochain
    // changement) rattrapera l'etat.
    private fun synchroniserAvecBackend() {
        val noms = repo.listerDossiersDesignes().map { it.nom }
        CoroutineScope(Dispatchers.IO).launch {
            try {
                ClovisApiClient(context).synchroniserDossiers(noms)
            } catch (e: Exception) {
                Log.w("DossiersPlugin", "Echec synchronisation dossiers designes avec le backend.", e)
            }
        }
    }

    override fun load() {
        super.load()
        synchroniserAvecBackend()
    }

    private fun DossierDesigne.toJson() = JSObject().apply {
        put("uri", uri.toString())
        put("nom", nom)
    }

    private fun ElementDossier.toJson() = JSObject().apply {
        put("uri", uri.toString())
        put("nom", nom)
        put("estDossier", estDossier)
        put("tailleOctets", tailleOctets)
    }

    @PluginMethod
    fun listerDossiersDesignes(call: PluginCall) {
        val tableau = JSArray()
        repo.listerDossiersDesignes().forEach { tableau.put(it.toJson()) }
        call.resolve(JSObject().put("dossiers", tableau))
    }

    /**
     * Ouvre le selecteur systeme (ACTION_OPEN_DOCUMENT_TREE). Le resultat
     * arrive dans handleChoixDossier ci-dessous (pattern Capacitor standard,
     * remplace rememberLauncherForActivityResult utilise cote Compose).
     */
    @PluginMethod
    fun choisirDossier(call: PluginCall) {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT_TREE)
        saveCall(call)
        startActivityForResult(call, intent, "handleChoixDossier")
    }

    @ActivityCallback
    private fun handleChoixDossier(call: PluginCall?, result: androidx.activity.result.ActivityResult) {
        val call = call ?: return
        val uri: Uri? = result.data?.data
        if (result.resultCode != android.app.Activity.RESULT_OK || uri == null) {
            call.reject("Aucun dossier choisi.")
            return
        }
        repo.ajouterDossierDesigne(uri)
        val dossier = repo.listerDossiersDesignes().firstOrNull { it.uri == uri }
        if (dossier == null) {
            call.reject("Dossier ajoute mais introuvable juste apres (cas inattendu).")
            return
        }
        synchroniserAvecBackend()
        call.resolve(dossier.toJson())
    }

    @PluginMethod
    fun retirerDossierDesigne(call: PluginCall) {
        val uri = call.getString("uri")
        if (uri == null) {
            call.reject("Parametre 'uri' manquant.")
            return
        }
        repo.retirerDossierDesigne(Uri.parse(uri))
        synchroniserAvecBackend()
        call.resolve()
    }

    @PluginMethod
    fun listerContenu(call: PluginCall) {
        val uri = call.getString("uri")
        if (uri == null) {
            call.reject("Parametre 'uri' manquant.")
            return
        }
        val tableau = JSArray()
        repo.listerContenu(Uri.parse(uri)).forEach { tableau.put(it.toJson()) }
        call.resolve(JSObject().put("elements", tableau))
    }

    // Ajoute le 30/08/2026 (correctif Claude chat) : lire_fichier /
    // chercher_par_contenu (core/exploration_dossier_mobile.py cote
    // clovis-backend) envoyaient deja cette question au telephone via le
    // canal temps reel, mais aucune methode native ne savait y repondre --
    // ce plugin s'arretait a lister le contenu, jamais a le lire.
    @PluginMethod
    fun lireFichier(call: PluginCall) {
        val uri = call.getString("uri")
        if (uri == null) {
            call.reject("Parametre 'uri' manquant.")
            return
        }
        val lecture = repo.lireFichier(Uri.parse(uri))
        if (lecture == null) {
            call.reject("Fichier introuvable ou illisible.")
            return
        }
        call.resolve(
            JSObject().apply {
                put("contenuBase64", lecture.contenuBase64)
                put("typeMime", lecture.typeMime)
                put("nomFichier", lecture.nomFichier)
                put("tailleOctets", lecture.tailleOctets)
            }
        )
    }

    @PluginMethod
    fun creerSousDossier(call: PluginCall) {
        val parentUri = call.getString("parentUri")
        val nom = call.getString("nom")
        if (parentUri == null || nom == null) {
            call.reject("Parametres 'parentUri' et 'nom' requis.")
            return
        }
        val ok = repo.creerSousDossier(Uri.parse(parentUri), nom)
        if (ok) call.resolve() else call.reject("Echec de la creation du sous-dossier.")
    }

    @PluginMethod
    fun creerFichier(call: PluginCall) {
        val parentUri = call.getString("parentUri")
        val nom = call.getString("nom")
        val typeMime = call.getString("typeMime") ?: "text/plain"
        if (parentUri == null || nom == null) {
            call.reject("Parametres 'parentUri' et 'nom' requis.")
            return
        }
        val ok = repo.creerFichier(Uri.parse(parentUri), nom, typeMime)
        if (ok) call.resolve() else call.reject("Echec de la creation du fichier.")
    }

    @PluginMethod
    fun renommer(call: PluginCall) {
        val elementUri = call.getString("elementUri")
        val nouveauNom = call.getString("nouveauNom")
        if (elementUri == null || nouveauNom == null) {
            call.reject("Parametres 'elementUri' et 'nouveauNom' requis.")
            return
        }
        val ok = repo.renommer(Uri.parse(elementUri), nouveauNom)
        if (ok) call.resolve() else call.reject("Echec du renommage.")
    }

    @PluginMethod
    fun supprimer(call: PluginCall) {
        val elementUri = call.getString("elementUri")
        if (elementUri == null) {
            call.reject("Parametre 'elementUri' manquant.")
            return
        }
        val ok = repo.supprimer(Uri.parse(elementUri))
        if (ok) call.resolve() else call.reject("Echec de la suppression.")
    }

    @PluginMethod
    fun deplacer(call: PluginCall) {
        val elementUri = call.getString("elementUri")
        val ancienParentUri = call.getString("ancienParentUri")
        val nouveauParentUri = call.getString("nouveauParentUri")
        if (elementUri == null || ancienParentUri == null || nouveauParentUri == null) {
            call.reject("Parametres 'elementUri', 'ancienParentUri' et 'nouveauParentUri' requis.")
            return
        }
        val ok = repo.deplacer(Uri.parse(elementUri), Uri.parse(ancienParentUri), Uri.parse(nouveauParentUri))
        if (ok) call.resolve() else call.reject("Echec du deplacement.")
    }
}

File: ios/App/App/dossiers/DossiersPlugin.swift
// Cree le 25/08/2026, Bourama : Lot 3B Partie 3 mobile (fusion Capacitor), iOS.
// Porte depuis clovis-mobile/ios-legacy-natif (Lot 2, DossiersScreen.swift
// + DossiersDesignesRepository.swift, copie telle quelle a cote de ce
// fichier) : meme logique (security-scoped bookmarks), l'ecran SwiftUI est
// retire, le UIDocumentPickerViewController est presente directement en
// UIKit depuis ce plugin (pas d'equivalent SwiftUI natif, meme raison
// documentee dans le fichier legacy).
//
// Cote JS (identique Android, mais uri devient un id de bookmark base64,
// PAS un chemin de fichier reel -- ne jamais l'afficher/interpreter cote
// web, le passer tel quel aux autres methodes de ce plugin) :
//   const Dossiers = registerPlugin<any>('Dossiers');
//   await Dossiers.choisirDossier();
//   await Dossiers.listerDossiersDesignes();
//   await Dossiers.listerContenu({ uri });
//   await Dossiers.creerSousDossier({ parentUri, nom });
//   await Dossiers.creerFichier({ parentUri, nom });
//   await Dossiers.renommer({ elementUri, nouveauNom });
//   await Dossiers.supprimer({ elementUri });
//   await Dossiers.deplacer({ elementUri, nouveauParentUri });
//   await Dossiers.retirerDossierDesigne({ uri });
import Foundation
import Capacitor
import UIKit
import UniformTypeIdentifiers

@objc(DossiersPlugin)
public class DossiersPlugin: CAPPlugin, CAPBridgedPlugin, UIDocumentPickerDelegate {
    public let identifier = "DossiersPlugin"
    public let jsName = "Dossiers"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "choisirDossier", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "listerDossiersDesignes", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "retirerDossierDesigne", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "listerContenu", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "creerSousDossier", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "creerFichier", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "renommer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "supprimer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deplacer", returnType: CAPPluginReturnPromise)
    ]

    // Ajoute le 26/08/2026 : miroir cote backend (voir
    // clovis-backend/core/dossiers_designes_mobile.py), meme role que cote
    // Android -- appele a l'ouverture de l'app (load()) et apres CHAQUE
    // changement, jamais bloquant (echec silencieux, log seulement).
    public override func load() {
        super.load()
        synchroniserAvecBackend()
    }

    private func synchroniserAvecBackend() {
        let noms = DossiersDesignesRepository.listerDossiersDesignes().map { $0.url.lastPathComponent }
        Task {
            try? await ClovisApiClient.synchroniserDossiers(noms)
        }
    }

    // Appel JS en attente pendant que le picker systeme est affiche (un seul
    // a la fois, coherent avec le fait que choisirDossier() bloque le JS
    // jusqu'a resolution de toute facon).
    private var appelChoixEnCours: CAPPluginCall?

    // Retrouve un dossier designe/element par son id (bookmark base64) parmi
    // les dossiers deja enregistres -- ne fonctionne QUE pour les dossiers
    // eux-memes (racine designee), pas pour un sous-element (voir
    // resoudreUrlElement pour le cas general, base sur le chemin).
    private func dossierDesigne(id: String) -> DossierDesigne? {
        DossiersDesignesRepository.listerDossiersDesignes().first { $0.id == id }
    }

    private func elementJson(_ e: ElementDossier) -> JSObject {
        JSObject([
            "uri": e.url.path,
            "nom": e.nom,
            "estDossier": e.estDossier,
            "tailleOctets": e.tailleOctets
        ])
    }

    @objc func choisirDossier(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let racine = self.bridge?.viewController else {
                call.reject("Impossible d'ouvrir le selecteur (pas de fenetre active).")
                return
            }
            self.appelChoixEnCours = call
            let picker = UIDocumentPickerViewController(forOpeningContentTypes: [.folder])
            picker.delegate = self
            racine.present(picker, animated: true)
        }
    }

    public func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        guard let call = appelChoixEnCours, let url = urls.first else {
            appelChoixEnCours?.reject("Aucun dossier choisi.")
            appelChoixEnCours = nil
            return
        }
        DossiersDesignesRepository.ajouterDossierDesigne(url: url)
        if let dossier = DossiersDesignesRepository.listerDossiersDesignes().first(where: { $0.url == url }) {
            synchroniserAvecBackend()
            call.resolve(["uri": dossier.id, "nom": dossier.url.lastPathComponent])
        } else {
            call.reject("Dossier ajoute mais introuvable juste apres (cas inattendu).")
        }
        appelChoixEnCours = nil
    }

    public func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
        appelChoixEnCours?.reject("Selection annulee.")
        appelChoixEnCours = nil
    }

    @objc func listerDossiersDesignes(_ call: CAPPluginCall) {
        let tableau = DossiersDesignesRepository.listerDossiersDesignes().map {
            ["uri": $0.id, "nom": $0.url.lastPathComponent] as [String: Any]
        }
        call.resolve(["dossiers": tableau])
    }

    @objc func retirerDossierDesigne(_ call: CAPPluginCall) {
        guard let uri = call.getString("uri") else {
            call.reject("Parametre 'uri' manquant.")
            return
        }
        DossiersDesignesRepository.retirerDossierDesigne(id: uri)
        synchroniserAvecBackend()
        call.resolve()
    }

    // 'uri' est soit un id de bookmark (dossier designe = racine), soit un
    // chemin de fichier reel deja resolu (sous-element, renvoye par
    // listerContenu ci-dessous) -- on tente les deux dans cet ordre.
    private func resoudreUrl(_ uri: String) -> URL? {
        if let dossier = dossierDesigne(id: uri) { return dossier.url }
        return URL(fileURLWithPath: uri)
    }

    @objc func listerContenu(_ call: CAPPluginCall) {
        guard let uri = call.getString("uri"), let url = resoudreUrl(uri) else {
            call.reject("Parametre 'uri' manquant ou invalide.")
            return
        }
        let elements = DossiersDesignesRepository.listerContenu(url).map { elementJson($0) }
        call.resolve(["elements": elements])
    }

    @objc func creerSousDossier(_ call: CAPPluginCall) {
        guard let parentUri = call.getString("parentUri"), let nom = call.getString("nom"),
              let parent = resoudreUrl(parentUri) else {
            call.reject("Parametres 'parentUri' et 'nom' requis.")
            return
        }
        if DossiersDesignesRepository.creerSousDossier(dansParent: parent, nom: nom) {
            call.resolve()
        } else {
            call.reject("Echec de la creation du sous-dossier.")
        }
    }

    @objc func creerFichier(_ call: CAPPluginCall) {
        guard let parentUri = call.getString("parentUri"), let nom = call.getString("nom"),
              let parent = resoudreUrl(parentUri) else {
            call.reject("Parametres 'parentUri' et 'nom' requis.")
            return
        }
        if DossiersDesignesRepository.creerFichier(dansParent: parent, nom: nom) {
            call.resolve()
        } else {
            call.reject("Echec de la creation du fichier.")
        }
    }

    @objc func renommer(_ call: CAPPluginCall) {
        guard let elementUri = call.getString("elementUri"), let nouveauNom = call.getString("nouveauNom"),
              let element = resoudreUrl(elementUri) else {
            call.reject("Parametres 'elementUri' et 'nouveauNom' requis.")
            return
        }
        if DossiersDesignesRepository.renommer(element, nouveauNom: nouveauNom) {
            call.resolve()
        } else {
            call.reject("Echec du renommage.")
        }
    }

    @objc func supprimer(_ call: CAPPluginCall) {
        guard let elementUri = call.getString("elementUri"), let element = resoudreUrl(elementUri) else {
            call.reject("Parametre 'elementUri' manquant.")
            return
        }
        if DossiersDesignesRepository.supprimer(element) {
            call.resolve()
        } else {
            call.reject("Echec de la suppression.")
        }
    }

    @objc func deplacer(_ call: CAPPluginCall) {
        guard let elementUri = call.getString("elementUri"), let nouveauParentUri = call.getString("nouveauParentUri"),
              let element = resoudreUrl(elementUri), let nouveauParent = resoudreUrl(nouveauParentUri) else {
            call.reject("Parametres 'elementUri' et 'nouveauParentUri' requis.")
            return
        }
        if DossiersDesignesRepository.deplacer(element, versParent: nouveauParent) {
            call.resolve()
        } else {
            call.reject("Echec du deplacement.")
        }
    }
}
